overlord/snapstate/backend/kernel_modules_components.go

```diff
@@ -0,0 +1,217 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package backend
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/progress"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/snap/naming"
+	"github.com/snapcore/snapd/systemd"
+)
+
+type NoKernelDriversError struct {
+	cref          naming.ComponentRef
+	kernelVersion string
+}
+
+func (e NoKernelDriversError) Error() string {
+	return fmt.Sprintf("%s does not contain firmware or components for %s",
+		e.cref, e.kernelVersion)
+}
+
+func cleanupMount(mountDir string, meter progress.Meter) error {
+	mountDir = filepath.Join(dirs.GlobalRootDir, mountDir)
+	// this also ensures that the mount unit stops
+	if err := removeMountUnit(mountDir, meter); err != nil {
+		return err
+	}
+
+	if err := os.RemoveAll(mountDir); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+type kernelModulesCleanupParts struct {
+	compMountDir    string
+	modulesMountDir string
+	rerunDepmod     bool
+}
+
+func cleanupKernelModulesSetup(parts *kernelModulesCleanupParts, kernelVersion string, meter progress.Meter) error {
+	if parts.modulesMountDir != "" {
+		if err := cleanupMount(parts.modulesMountDir, meter); err != nil {
+			return err
+		}
+		if parts.rerunDepmod {
+			if err := runDepmod("/usr", kernelVersion); err != nil {
+				return err
+			}
+		}
+	}
+
+	if parts.compMountDir != "" {
+		if err := cleanupMount(parts.compMountDir, meter); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func checkKernelModulesCompContent(mountDir, kernelVersion string) (bool, bool) {
+	hasModules := osutil.IsDirectory(filepath.Join(mountDir, "modules", kernelVersion))
+	hasFirmware := osutil.IsDirectory(filepath.Join(mountDir, "firmware"))
+	return hasModules, hasFirmware
+}
+
+func componentMountPoint(componentName, kernelVersion string) string {
+	return filepath.Join("/run/mnt/kernel-modules/", kernelVersion, componentName)
+}
+
+func modulesMountPoint(componentName, kernelVersion string) string {
+	return filepath.Join("/usr/lib/modules", kernelVersion, "updates", componentName)
+}
+
+// SetupKernelModulesComponent creates and starts mount units for
+// kernel-modules components.
+func (b Backend) SetupKernelModulesComponent(cpi snap.ContainerPlaceInfo, cref naming.ComponentRef, kernelVersion string, meter progress.Meter) (err error) {
+	var sysd systemd.Systemd
+	if b.preseed {
+		sysd = systemd.NewEmulationMode(dirs.GlobalRootDir)
+	} else {
+		sysd = systemd.New(systemd.SystemMode, meter)
+	}
+
+	// Restore state if something goes wrong
+	var cleanOnFailure kernelModulesCleanupParts
+	defer func() {
+		if err == nil {
+			return
+		}
+		if err := cleanupKernelModulesSetup(&cleanOnFailure, kernelVersion, meter); err != nil {
+			logger.Noticef("while cleaning up a failed kernel-modules set-up: %v", err)
+		}
+	}()
```

Comment:
should we keep doing this here? or should we make it a problem of the caller, which might decide to do a switch back instead?

Reply:
That's a good point as indeed in some cases we want to restore the old state and we might not want to do previously a clean-up.